### ISSUE TITLE
Add const Keyword to the C bindings

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#46]
**Originally by _erezh_ on 2008-10-16 16:15:34 -0500**

---
# Add const keyword to the C bindingsworkgroup list: [mailto:mpi-const@lists.mpi-forum.org]
## April meeting straw votes (4/29/2008)

Yes: 18[[BR]]
No: 3[[BR]]
Abstain: 9[[BR]]
## Background:

The const keyword in C defines a contract between the library implementer and the library user. Using the const keyword the library contracts that it will not change its input object. This contract enables some compile-time optimization, but more important it provides clearer and const-correct interface to the library user. (more on [http://en.wikipedia.org/wiki/Const_correctness])

The MPI C bindings as defined by the MPI 1.1 & 2.0 standards are missing the const keyword for many of the input only parameters.
## Proposal:

Add the const keyword to the API’s listed below.

-Note: adding the const keyword to the point-to-point and collectives API’s send buffer depends on the the proposal to ticket #45 Remove Send Buffer Access Restriction*
## Rational:

The missing const keyword means that the contract between the user and the MPI library is weaker than it should be. Users need to cast away const-ness before calling MPI.
The C++ binding has already implemented a const correct interface. This change catches up with the C++ interface. (note that the C++ interface implementations cast away the const-ness when calling their C binding).
## Impact on existing implementations:

MPI library implementations need to change their C header files.
## Impact on applications:

Risk: compilers that still do not support the const keyword. (i.e., compilers that are not ANSI/ISO compliant)

Backward compatibility: There should not be any backward compatibility issue. Already compiled programs should run without a problem with a new dynamically loaded library as there is no change to the size of the parameters. The C **linkers** do not check for const correctness, thus no error here. Recompiled programs should see no new compilation errors or warning since the const keyword guarantees a stronger contract; any non const pointer is automatically promoted to be const.
## Prototype

This proposal was tested using Microsoft Visual Studio, GNU, Intel and Pathscale compilers using the MPICH2 and the Intel test suites. All tests compiled without any extra warning or errors and run correctly.
## Adding the const keyword:

A compiled list of API’s that are missing the const keywords is listed below. The API's are grouped by functionality and the parameter where the const keyword is applied. The last section lists 4 API’s where this proposal suggests not to add the const keyword as it would break source level backward compatibility.

```
#!c
/* Chapter 3: Point-to-point send buffer */
int MPI_Bsend(const void*, int, MPI_Datatype, int, int, MPI_Comm);
int MPI_Bsend_init(const void*, int, MPI_Datatype, int,int, MPI_Comm, MPI_Request*);
int MPI_Ibsend(const void*, int, MPI_Datatype, int, int, MPI_Comm, MPI_Request*);
int MPI_Irsend(const void*, int, MPI_Datatype, int, int, MPI_Comm, MPI_Request*);
int MPI_Isend(const void*, int, MPI_Datatype, int, int, MPI_Comm, MPI_Request*);
int MPI_Issend(const void*, int, MPI_Datatype, int, int, MPI_Comm, MPI_Request*);
int MPI_Rsend(const void*, int, MPI_Datatype, int, int, MPI_Comm);
int MPI_Rsend_init(const void*, int, MPI_Datatype, int,int, MPI_Comm, MPI_Request*);
int MPI_Send(const void*, int, MPI_Datatype, int, int, MPI_Comm);
int MPI_Send_init(const void*, int, MPI_Datatype, int, int, MPI_Comm, MPI_Request*);
int MPI_Sendrecv(const void*, int, MPI_Datatype,int, int, void*, int, MPI_Datatype, int, int, MPI_Comm, MPI_Status*);
int MPI_Ssend(const void*, int, MPI_Datatype, int, int, MPI_Comm);
int MPI_Ssend_init(const void*, int, MPI_Datatype, int,int, MPI_Comm, MPI_Request*);

/* Chapter 3: Miscellany */
int MPI_Test_cancelled(const MPI_Status*, int*);

/* Chapter 4: Index/Count arrays */
int MPI_Type_create_darray(int, int, int, const int [], const int [], const int [], const int [], int, MPI_Datatype, MPI_Datatype*);
int MPI_Type_create_hindexed(int, const int [], const MPI_Aint [], MPI_Datatype, MPI_Datatype*);
int MPI_Type_create_indexed_block(int, int, const int [], MPI_Datatype, MPI_Datatype*);
int MPI_Type_create_struct(int, const int [], const MPI_Aint [], const MPI_Datatype [], MPI_Datatype*);
int MPI_Type_create_subarray(int, const int [], const int [], const int [], int, MPI_Datatype, MPI_Datatype*);
int MPI_Type_hindexed(int, const int*, const MPI_Aint*, MPI_Datatype, MPI_Datatype*);
int MPI_Type_indexed(int, const int*, const int*, MPI_Datatype, MPI_Datatype*);
int MPI_Type_struct(int, const int*, const MPI_Aint*, const MPI_Datatype*, MPI_Datatype*);

/* Chapter 4: Pack/Unpack datarep string and input buffer */
int MPI_Pack(const void*, int, MPI_Datatype, void*, int, int*,  MPI_Comm);
int MPI_Pack_external(const char*, const void*, int, MPI_Datatype, void*, MPI_Aint, MPI_Aint*); 
int MPI_Pack_external_size(const char*, int, MPI_Datatype, MPI_Aint*); 
int MPI_Unpack(const void*, int, const int*, void*, int, MPI_Datatype, MPI_Comm);
int MPI_Unpack_external(const char*, const void*, MPI_Aint, MPI_Aint*, void*, int, MPI_Datatype); 

/* Chapter 4: Miscellany */
int MPI_Get_address(const void*, MPI_Aint*);

/* Chapter 5: Collectives send buffer */
int MPI_Allgather(const void* , int, MPI_Datatype, void*, int, MPI_Datatype, MPI_Comm);
int MPI_Allgatherv(const void* , int, MPI_Datatype, void*, const int*, const int*, MPI_Datatype, MPI_Comm);
int MPI_Allreduce(const void* , void*, int, MPI_Datatype, MPI_Op, MPI_Comm);
int MPI_Alltoall(const void* , int, MPI_Datatype, void*, int, MPI_Datatype, MPI_Comm);
int MPI_Alltoallv(const void* , const int*, const int*, MPI_Datatype, void*, const int*, const int*, MPI_Datatype, MPI_Comm);
int MPI_Alltoallw(const void*, const int [], const int [], const MPI_Datatype [], void*, const int [], const int [], const MPI_Datatype [], MPI_Comm);
int MPI_Exscan(const void*, void*, int, MPI_Datatype, MPI_Op, MPI_Comm) ;
int MPI_Gather(const void* , int, MPI_Datatype, void*, int, MPI_Datatype, int, MPI_Comm); 
int MPI_Gatherv(const void* , int, MPI_Datatype, void*, const int*, const int*, MPI_Datatype, int, MPI_Comm); 
int MPI_Reduce(const void* , void*, int, MPI_Datatype, MPI_Op, int, MPI_Comm);
int MPI_Reduce_scatter(const void* , void*, const int*, MPI_Datatype, MPI_Op, MPI_Comm);
int MPI_Scan(const void* , void*, int, MPI_Datatype, MPI_Op, MPI_Comm );
int MPI_Scatter(const void* , int, MPI_Datatype, void*, int, MPI_Datatype, int, MPI_Comm);
int MPI_Scatterv(const void* , const int*, const int*,  MPI_Datatype, void*, int, MPI_Datatype, int, MPI_Comm);

/* Chapter 6: Index/Count arrays */
int MPI_Group_excl(MPI_Group, int, const int*, MPI_Group*);
int MPI_Group_incl(MPI_Group, int, const int*, MPI_Group*);
int MPI_Group_translate_ranks(MPI_Group, int, const int*, MPI_Group, int*);

/* Chapter 6: Input strings */
int MPI_Comm_set_name(MPI_Comm, const char*);
int MPI_Type_set_name(MPI_Datatype, const char*);
int MPI_Win_set_name(MPI_Win, const char*);

/* Chapter 7: Index/Count arrays */
int MPI_Cart_create(MPI_Comm, int, const int*, const int*, int, MPI_Comm*);
int MPI_Cart_map(MPI_Comm, int, const int*, const int*, const int*);
int MPI_Cart_rank(MPI_Comm, const int*, int*);
int MPI_Cart_sub(MPI_Comm, const int*, MPI_Comm*);

int MPI_Graph_create(MPI_Comm, int, const int*, const int*, int, MPI_Comm*);
int MPI_Graph_map(MPI_Comm, int, const int*, const int*, int*);

/* Chapter 8: Input strings */
int MPI_Add_error_string(int, const char*);

/* Chapter 9: Input strings */
int MPI_Info_delete(MPI_Info, const char*);
int MPI_Info_get(MPI_Info, const char*, int, char*, int*);
int MPI_Info_get_valuelen(MPI_Info, const char*, int*, int*);
int MPI_Info_set(MPI_Info, const char*, const char*);

/* Chapter 10: Input strings */
int MPI_Close_port(const char*);
int MPI_Comm_accept(const char*, MPI_Info, int, MPI_Comm, MPI_Comm*);
int MPI_Comm_connect(const char*, MPI_Info, int, MPI_Comm, MPI_Comm*);
int MPI_Comm_spawn(const char*, char*[], int, MPI_Info, int, MPI_Comm, MPI_Comm*, int []);
int MPI_Lookup_name(const char*, MPI_Info, char*);
int MPI_Publish_name(const char*, MPI_Info, const char*);
int MPI_Unpublish_name(const char*, MPI_Info, const char*);

/* Chapter 11: RMA buffer */
int MPI_Accumulate(const void*, int, MPI_Datatype, int, MPI_Aint, int, MPI_Datatype,  MPI_Op, MPI_Win);
int MPI_Put(const void*, int, MPI_Datatype, int, MPI_Aint, int, MPI_Datatype, MPI_Win);

/* Chapter 13: File IO write buffer */
int MPI_File_iwrite(MPI_File, const void*, int, MPI_Datatype, MPI_Request*);
int MPI_File_iwrite_at(MPI_File, MPI_Offset, const void*, int, MPI_Datatype, MPI_Request*);
int MPI_File_iwrite_shared(MPI_File, const void*, int, MPI_Datatype, MPI_Request*);
int MPI_File_write(MPI_File, const void*, int, MPI_Datatype, MPI_Status*);
int MPI_File_write_all(MPI_File, const void*, int, MPI_Datatype, MPI_Status*);
int MPI_File_write_all_begin(MPI_File, const void*, int, MPI_Datatype);
int MPI_File_write_all_end(MPI_File, const void*, MPI_Status*);
int MPI_File_write_at(MPI_File, MPI_Offset, const void*, int, MPI_Datatype, MPI_Status*);
int MPI_File_write_at_all(MPI_File, MPI_Offset, const void*, int, MPI_Datatype, MPI_Status*);
int MPI_File_write_at_all_begin(MPI_File, MPI_Offset, const void*, int, MPI_Datatype);
int MPI_File_write_at_all_end(MPI_File, const void*, MPI_Status*);
int MPI_File_write_ordered(MPI_File, const void*, int, MPI_Datatype, MPI_Status*);
int MPI_File_write_ordered_begin(MPI_File, const void*, int, MPI_Datatype);
int MPI_File_write_ordered_end(MPI_File, const void*, MPI_Status*);
int MPI_File_write_shared(MPI_File, const void*, int, MPI_Datatype, MPI_Status*);

/* Chapter 13: Input strings */
int MPI_File_delete(const char*, MPI_Info);
int MPI_File_open(MPI_Comm, const char*, int, MPI_Info, MPI_File*);
int MPI_File_set_view(MPI_File, MPI_Offset, MPI_Datatype, MPI_Datatype, const char*, MPI_Info);

/* Chapter 15: Miscellany */
int MPI_Address(const void*, MPI_Aint*);

/* Chapter 16: Miscellany */
int MPI_Status_c2f(const MPI_Status*, MPI_Fint*);
int MPI_Status_f2c(const MPI_Fint*, MPI_Status*);

/* NOT CHANGING the following API's to keep backward compatibility and ease of use */
int MPI_Comm_spawn(const char*, /*const*/char*[], int, MPI_Info, int, MPI_Comm, MPI_Comm*, int []);
int MPI_Comm_spawn_multiple(int, /*const*/char*[], /*const*/char**[], /*const*/int [],/*const*/MPI_Info [], int, MPI_Comm, MPI_Comm*, int []); 
int MPI_Group_range_excl(MPI_Group, int, /*const*/int [][3], MPI_Group*);
int MPI_Group_range_incl(MPI_Group, int, /*const*/int [][3], MPI_Group*);
```
## Entry to the Change Log

The 'const' keyword has been added to many functions in chapters 3-11, 13, 15, and 16.

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2008-10-26 09:40:06 -0500**

---

Changed back to had first reading, per Oct 2008 meeting.  Needs 4 reviewers

**Comment by mpiforumbot** [comment]
**Originally by _buntinas_ on 2008-12-05 16:05:32 -0600**

---

Darius will review.

**Comment by mpiforumbot** [comment]
**Originally by _asupalov_ on 2008-12-05 16:28:05 -0600**

---

See https://svn.mpi-forum.org/trac/mpi-forum-web/ticket/45#comment:7 . Since this matter is dependent on #45, this matter should be suspended until we have satisfactory resolution there.

**Comment by mpiforumbot** [comment]
**Originally by _asupalov_ on 2008-12-05 18:17:13 -0600**

---

[Adding a proper comment. Still learning this Wiki.]

See https://svn.mpi-forum.org/trac/mpi-forum-web/ticket/45#comment:7 . Since this matter is dependent on #45, this matter should be suspended until we have satisfactory resolution there.

**Comment by mpiforumbot** [comment]
**Originally by _rlgraham_ on 2008-12-13 17:21:54 -0600**

---

Reviewed as chapter author for chapters 3 and 4 - changes are fine.   Rich Graham

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2008-12-14 19:50:42 -0600**

---

Reviewed as chapter author for 7, 9, 11; seems correct functions have been identified. However, I'm not convinced what this proposal brings; also, if we start introducing constness, was it discussed which other things (count arguments, tags, ...) might be reasobanly const'ed?

Jesper

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2008-12-15 18:25:35 -0600**

---

Needs update to change log that lists which chapters (does not need sections, pages, or lines)

The const keyword has been added to many functions in chapters 3-11, 13, 15, and 16.

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2008-12-15 18:35:37 -0600**

---

Before the 2nd vote, the Forum requires these three steps:
1. Update the change log
2. Chapter authors need to review the changes (primarily for completeness)
3. An open source implementation must be posted

In addition, this change has a dependency on #45

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-01-12 12:50:21 -0600**

---

change log has been updated as above.

answer for Jesper above: yes count array etc have the const modifiers (e.g "const int \* counts", see MPI_Alltoallv above).
copy by value args (like 'int count') do not require the const modifier in C/C++ and have no semantic meaning for the user of the function. (and it looks unnatural to any C/C++ programmer)

**Comment by mpiforumbot** [comment]
**Originally by _rrtreu_ on 2009-01-14 14:03:29 -0600**

---

I have reviewed Chapter 6 and confirmed the plausible routines for "const" are all identified.  

This is not intended as a statement one way or the other about whether "const" should be added to the indicated parameters.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-01-15 14:48:02 -0600**

---

I reviewed the I/O Chapter (13), and it looks ok. MPI_Register_datarep seems to be missing though. Shouldn't it have a const char *datarep and const void *extra_state? The C++ binding is missing the const for extra_state.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-19 13:24:30 -0600**

---

The list is complete for all MPI functions in Chapter 5. However, what about the void *invec in the MPI_User_function on p168:29? 

Best,
  Torsten

**Comment by mpiforumbot** [comment]
**Originally by _dgsolt_ on 2009-01-21 12:03:54 -0600**

---

Chapter 10 looks good.  I don't understand why the 'int *' argument in MPI_Comm_spawn_multiple is different from the 'int *' argument in MPI_Alltoallv though.  MPI_Comm_spawn_multiple is listed in the "Not changed section", I don't understand why we don't change at least arguments #2 and #4:

```
int MPI_Comm_spawn_multiple(int, char*[], char**[],int [], MPI_Info [], int, MPI_Comm, MPI_Comm*, int []); 
```

to

```
int MPI_Comm_spawn_multiple(int, const char*[], char**[], const int [], MPI_Info [], int, MPI_Comm, MPI_Comm*, int []); 
```

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-02-06 18:33:26 -0600**

---

Replying to dgsolt:

thanks,

You can not really change argument #2 without breaking backward compatability. there is no implicit casting from char_[] to const char_[](notice that its a pointer to pointer).

However we could have changed param #4 (but not #9) to 'const int[]' (I think it was an oversight)

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-02-06 18:35:23 -0600**

---

Replying to rsthakur:

> I reviewed the I/O Chapter (13), and it looks ok. MPI_Register_datarep seems to be missing though. Shouldn't it have a const char *datarep and const void *extra_state? The C++ binding is missing the const for extra_state.

We should not add const to extra_state as we don't know the user intent.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-02-06 18:45:32 -0600**

---

Replying to erezh:

> We should not add const to extra_state as we don't know the user intent.

We don't know the user intent for extra_state, but we know that the function MPI_Register_datarep will not modify it.

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-02-10 11:31:13 -0600**

---

Replying to rsthakur:

> We don't know the user intent for extra_state, but we know that the function MPI_Register_datarep will not modify it.
> However the callback function does take extra_state as non-const which will force MPI_Register_datarep to cast the const away (unless we change the callback prototype, which will break backcompat)

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-02-10 11:36:49 -0600**

---

OK, but we need the const char *datarep, right?

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-02-10 11:47:15 -0600**

---

Replying to rsthakur:

> OK, but we need the const char *datarep, right?
> Yes.

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-02-11 12:09:25 -0600**

---

Restoring to the original descriptoin as voted on! (with the extra const in MPI_Unpack). Ticket #129 is posted to remove the extra const from MPI_Unpack in this proposal.

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-02-11 12:24:27 -0600**

---

ticket #130 adds two more functions to the list
update to MPI_Comm_spawn_multiple
update to MPI_Register_datarep

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-03-12 13:21:03 -0500**

---

Implementation is available from ANL

tarballs
http://www.mcs.anl.gov/research/projects/mpich2/downloads/tarballs/nightly/const

svn
https://svn.mcs.anl.gov/repos/mpi/mpich2/branches/dev/mpi_binding_const

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-03-12 14:36:57 -0500**

---

**Attachment added:** [`mpich2-const-r4035.diff`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/46/mpich2-const-r4035.diff) (478.7 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-03-12 14:38:46 -0500**

---

I've uploaded the diff compared to the mpich2 trunk (r4035). The full tarball corresponding to r4035 will be on the above site in a few minutes.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-04-04 12:37:31 -0500**

---

The proposal should clearly state, that the shown interfaces are not the future C-interfaces in MPI-2.2. The list only shows, on which arguments the const will be added. The argument names, as shown in MPI-2.1, should not be not removed.

How was it possible, that the list is wrong: See tickets #129 and #130.

I propose that the list of this ticket is modified according to the content of tickets #129 and #130 as soon as they are reviewed.

I propose to switch now to "author rework"
and to switch back to status "reviewed" as soon as #129 and #130 is included.
Additionally, the text of the "Proposal"

  Add the const keyword to the API’s listed below.

should be changed to

  Add the const keywords to the C-API's of MPI-2.1 according to the 
  positions shown in the API's listed below.

A ticket with well-known incorrect content cannot get a second vote.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-04-04 12:40:08 -0500**

---

Erez has merged all the tickets into #140. #140 is the new ticket for this.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-04-04 13:08:21 -0500**

---

The proposal for 
  int MPI_Comm_spawn(const char_, char_[], int, MPI_Info, int, MPI_Comm, MPI_Comm*, int []);
is wrong: see 297.10-16: [[BR]]
1st and 2nd argument should be const.
What are the rules for the 4th argument?

How can we get to a correct list?

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-04-04 13:31:22 -0500**

---

Replying to RolfRabenseifner:

> The proposal for 
>   int MPI_Comm_spawn(const char_, char_[], int, MPI_Info, int, MPI_Comm, MPI_Comm*, int []);
> is wrong: see 297.10-16: [[BR]]
> 1st and 2nd argument should be const.
> What are the rules for the 4th argument?
> The const keywork was explicitly not added to the second argument since it will break backward compatability.
> The 4th arg is a handle and thus does not require the const modifier.
> 
> How can we get to a correct list?
> The list has been reviewed and implemented several times with only 1 mistake in the above text, plus 3 additions that were not in the original text. the corrections are in tickets #129 and #130.  The merged text is in ticket #140, which I believe is the correct list. please review #140.

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-04-08 09:18:03 -0500**

---

this ticket was replaced by ticket #140
